lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -266,3 +266,100 @@ void ExecutionEngine::compile(Function *F, CompilationContext &cctx,
+    return true;
+  }
+  if (llvm::isa<Placeholder>(N)) {
+    return false;
+  }
+  for (unsigned idx = 0, e = N->getNumInputs(); idx < e; ++idx) {
+    auto input = N->getNthInput(idx);
+    if (!llvm::isa<Constant>(input) && !llvm::isa<SplatNode>(input)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/// Evaluates a provided constant operation \p C in the context of the function
+/// \p F.
+/// \returns a constant representing a result of computation.
+Node *evaluateConstantOperation(Function *F, Node *C) {
+  assert(isConstantOperation(C) && "Expected a constant expression");
+  // Create a temporary execution engine to compute only this specific
+  // operation.
+  PlaceholderBindings bindings;
+  ExecutionEngine EE("Interpreter");
+  Module &mod = *F->getParent();
+  // Create a temporary function to perform the constant operation.
+  Function *constEvaluationF = mod.createFunction(constEvaluationFunctionName);
+  unsigned constIdx = 0;
+  // Mapping from existing nodes to the new ones.
+  NodeMap currToNew;
+  // Clone the constant operation and some of its inputs if necessary.
+  auto *clonedC = recursiveClone(constEvaluationF, C, currToNew);
+  SaveNode *SN = constEvaluationF->createSave("save", clonedC);
+  auto *outputPH = SN->getPlaceholder();
+  bindings.allocate(outputPH);
+  // Run the temporary execution engine to perform this constant operation
+  // evaluation.
+  EE.compile(CompilationMode::Infer, constEvaluationF);
+  EE.run(bindings);
+  // Get the result of the constant operation and create a new constant from it.
+  Tensor *outputTensor = bindings.get(outputPH);
+  // Remove the temporary function.
+  mod.eraseFunction(constEvaluationF);
+  // Return the newly created constant, which is the result of a constant
+  // operation.
+  return mod.createConstant(strFormat("const_%u", constIdx++), *outputTensor);
+}
+} // namespace
+
+bool glow::constantFold(Function *F, CompilationContext &cctx) {
+  // Avoid recursively constant folding a function that is submitted for
+  // constant folding through a temporary execution engine.
+  if (F->getName().startswith(constEvaluationFunctionName)) {
+    return false;
+  }
+  LOG_SCOPE(F->getLogContext(), "glow::constantFold")
+  bool changed = false;
+  // Set of constant operations to be computed.
+  llvm::SmallVector<Node *, 64> worklist;
+  // Collect all non-trivial constant operations.
+  for (auto &N : F->getNodes()) {
+    // Skip trivial nodes/operations that do not require any constant
+    // computations.
+    if (llvm::isa<Storage>(N) || llvm::isa<Constant>(N) ||
+        llvm::isa<SplatNode>(N)) {
+      continue;
+    }
+    if (isConstantOperation(&N)) {
+      worklist.push_back(&N);
+    }
+  }
+  // Compute the result of each collected non-trivial constant operation.
+  while (!worklist.empty()) {
+    auto *C = worklist.pop_back_val();
+    assert(C->getNumResults() == 1 && "Expected a single result");
```

Comment:
Could remove this limitation of needing an op with a single result by returning a vector of Constants corresponding to each of the results of the op that is being constant folded.

Reply:
Fixed.